ops/pebble.py

```diff
@@ -752,5 +861,198 @@ def get_services(self, names: List[str] = None) -> List[ServiceInfo]:
         query = None
         if names is not None:
             query = {'names': ','.join(names)}
-        result = self._request('GET', '/v1/services', query)
-        return [ServiceInfo.from_dict(info) for info in result['result']]
+        resp = self._request('GET', '/v1/services', query)
+        return [ServiceInfo.from_dict(info) for info in resp['result']]
+
+    def read_file(self, path: str, destination: BinaryIO):
+        """Read a file from the remote system and write content to destination."""
+        query = {
+            'action': 'read',
+            'path': path,
+        }
+        headers = {'Accept': 'multipart/form-data'}
+        response = self._request_raw('GET', '/v1/files', query, headers)
+        resp = self._parse_read_multipart(response, {path: destination})
+        self._raise_on_path_error(resp, path)
+
+    @staticmethod
+    def _raise_on_path_error(resp, path):
+        result = resp['result'] or []  # in case it's null instead of []
+        paths = {item['path']: item for item in result}
+        if path not in paths:
+            raise ProtocolError('path not found in response metadata: {}'.format(resp))
+        error = paths[path].get('error')
+        if error:
+            raise PathError(error['kind'], error['message'])
+
+    @classmethod
+    def _parse_read_multipart(cls, response: http.client.HTTPResponse,
+                              destinations: Dict[str, BinaryIO]) -> Dict:
+        """Parse a multipart HTTP response from the read-files API.
+
+        Return "response" metadata field decoded from JSON, and write content
+        to file-like object in destinations dictionary (keyed by path).
+        Currently the content is entirely loaded into memory, but the goal is
+        to stream that in future (the signature of read_file won't change).
+        """
+        options = cls._parse_content_type(response.headers, 'multipart/form-data')
+        boundary = options.get('boundary', '')
+        if len(boundary) < 32:
```

Comment:
Where does 32 come from? Is that part of the spec? Or it's a reasonable minimum to test against for a coarse grained check?

Reply:
It was copied from the `minBoundaryLength` in the server. However, per John, I've just avoided the length check altogether and only check whether it's present or not.